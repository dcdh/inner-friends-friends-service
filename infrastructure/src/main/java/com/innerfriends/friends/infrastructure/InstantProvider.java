package com.innerfriends.friends.infrastructure;

import javax.enterprise.context.ApplicationScoped;
import java.time.Instant;

@ApplicationScoped
public class InstantProvider {

    public Instant now() {
        return Instant.now();
    }

}
