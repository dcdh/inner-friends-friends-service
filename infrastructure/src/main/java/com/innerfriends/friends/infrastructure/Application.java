package com.innerfriends.friends.infrastructure;

import com.innerfriends.friends.domain.*;
import com.innerfriends.friends.domain.usecase.*;

import javax.enterprise.context.ApplicationScoped;
import javax.enterprise.inject.Produces;
import java.time.LocalDateTime;
import java.time.ZoneOffset;
import java.util.UUID;

public class Application {

    @Produces
    @ApplicationScoped
    public EstablishAFriendshipFromFriendWithToFriendUseCase establishAFriendshipFromFriendWithToFriendUseCaseProducer(final FriendRepository friendRepository) {
        return new EstablishAFriendshipFromFriendWithToFriendUseCase(friendRepository);
    }

    @Produces
    @ApplicationScoped
    public EstablishAFriendshipToFriendWithFromFriendUseCase establishAFriendshipToFriendWithFromFriendUseCaseProducer(final InvitationCodeGeneratedRepository invitationCodeGeneratedRepository,
                                                                                                                       final FriendRepository friendRepository,
                                                                                                                       final LocalDateTimeProvider localDateTimeProvider) {
        return new EstablishAFriendshipToFriendWithFromFriendUseCase(invitationCodeGeneratedRepository, friendRepository, localDateTimeProvider);
    }

    @Produces
    @ApplicationScoped
    public GenerateInvitationCodeUseCase generateInvitationCodeUseCaseProducer(final InvitationCodeGeneratedRepository invitationCodeGeneratedRepository,
                                                                               final InvitationCodeGenerator invitationCodeGenerator,
                                                                               final GeneratedAtProvider generatedAtProvider) {
        return new GenerateInvitationCodeUseCase(invitationCodeGeneratedRepository, invitationCodeGenerator, generatedAtProvider);
    }

    @Produces
    @ApplicationScoped
    public GetFriendUseCase getFriendUseCaseProducer(final FriendRepository friendRepository) {
        return new GetFriendUseCase(friendRepository);
    }

    @Produces
    @ApplicationScoped
    public GetFriendOfFriendUseCase getFriendOfFriendUseCaseProducer(final FriendRepository friendRepository) {
        return new GetFriendOfFriendUseCase(friendRepository);
    }

    @Produces
    @ApplicationScoped
    public GetInFriendshipWithUseCase getInFriendshipWithUseCaseProducer(final FriendRepository friendRepository) {
        return new GetInFriendshipWithUseCase(friendRepository);
    }

    @Produces
    @ApplicationScoped
    public GetInFriendshipWithMutualFriendsUseCase getInFriendshipWithMutualFriendsUseCaseProducer(final MutualFriendsRepository mutualFriendsRepository) {
        return new GetInFriendshipWithMutualFriendsUseCase(mutualFriendsRepository);
    }

    @Produces
    @ApplicationScoped
    public GetFriendsOfFriendMutualFriendsUseCase getFriendsOfFriendMutualFriendsUseCaseProducer(final MutualFriendsRepository mutualFriendsRepository) {
        return new GetFriendsOfFriendMutualFriendsUseCase(mutualFriendsRepository);
    }

    @Produces
    @ApplicationScoped
    public RegisterNewFriendIntoThePlatformUseCase registerNewFriendIntoThePlatformUseCaseProducer(final FriendRepository friendRepository) {
        return new RegisterNewFriendIntoThePlatformUseCase(friendRepository);
    }

    @Produces
    @ApplicationScoped
    public WriteBioUseCase writeBioUseCaseProducer(final FriendRepository friendRepository) {
        return new WriteBioUseCase(friendRepository);
    }

    @Produces
    @ApplicationScoped
    public GeneratedAtProvider generatedAtProvider() {
        return new GeneratedAtProvider() {
            @Override
            public GeneratedAt now() {
                return new GeneratedAt(LocalDateTime.now(ZoneOffset.UTC));
            }
        };
    }

    @Produces
    @ApplicationScoped
    public LocalDateTimeProvider localDateTimeProvider() {
        return new LocalDateTimeProvider() {
            @Override
            public LocalDateTime now() {
                return LocalDateTime.now(ZoneOffset.UTC);
            }
        };
    }

    @Produces
    @ApplicationScoped
    public InvitationCodeGenerator invitationCodeGenerator() {
        return new InvitationCodeGenerator() {
            @Override
            public InvitationCode generate(final FromFriendId fromFriendId) {
                return new InvitationCode(UUID.randomUUID());
            }
        };
    }

}
