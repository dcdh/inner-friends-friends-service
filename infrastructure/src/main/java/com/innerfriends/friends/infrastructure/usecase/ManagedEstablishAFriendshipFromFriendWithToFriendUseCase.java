package com.innerfriends.friends.infrastructure.usecase;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.innerfriends.friends.domain.Friend;
import com.innerfriends.friends.domain.UseCase;
import com.innerfriends.friends.domain.usecase.EstablishAFriendshipFromFriendWithToFriendCommand;
import com.innerfriends.friends.domain.usecase.EstablishAFriendshipFromFriendWithToFriendUseCase;
import com.innerfriends.friends.infrastructure.InstantProvider;
import com.innerfriends.friends.infrastructure.outbox.FromFriendEstablishedAFriendshipWithToFriendEvent;
import io.debezium.outbox.quarkus.ExportedEvent;

import javax.enterprise.context.ApplicationScoped;
import javax.enterprise.event.Event;
import javax.transaction.Transactional;
import java.util.Objects;

@ApplicationScoped
public class ManagedEstablishAFriendshipFromFriendWithToFriendUseCase implements UseCase<Friend, EstablishAFriendshipFromFriendWithToFriendCommand> {

    private final EstablishAFriendshipFromFriendWithToFriendUseCase establishAFriendshipFromFriendWithToFriendUseCase;
    private final Event<ExportedEvent<?, ?>> event;
    private final ObjectMapper objectMapper;
    private final InstantProvider instantProvider;

    public ManagedEstablishAFriendshipFromFriendWithToFriendUseCase(final EstablishAFriendshipFromFriendWithToFriendUseCase establishAFriendshipFromFriendWithToFriendUseCase,
                                                                    final Event<ExportedEvent<?, ?>> event,
                                                                    final ObjectMapper objectMapper,
                                                                    final InstantProvider instantProvider) {
        this.establishAFriendshipFromFriendWithToFriendUseCase = Objects.requireNonNull(establishAFriendshipFromFriendWithToFriendUseCase);
        this.event = Objects.requireNonNull(event);
        this.objectMapper = Objects.requireNonNull(objectMapper);
        this.instantProvider = Objects.requireNonNull(instantProvider);
    }

    @Transactional
    @Override
    public Friend execute(final EstablishAFriendshipFromFriendWithToFriendCommand command) {
        final Friend friendConnected = establishAFriendshipFromFriendWithToFriendUseCase.execute(command);
        event.fire(FromFriendEstablishedAFriendshipWithToFriendEvent.of(friendConnected, objectMapper, instantProvider));
        return friendConnected;
    }
}
