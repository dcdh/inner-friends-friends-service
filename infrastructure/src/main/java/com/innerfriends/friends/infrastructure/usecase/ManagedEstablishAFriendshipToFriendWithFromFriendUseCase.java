package com.innerfriends.friends.infrastructure.usecase;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.innerfriends.friends.domain.Friend;
import com.innerfriends.friends.domain.UseCase;
import com.innerfriends.friends.domain.usecase.EstablishAFriendshipToFriendWithFromFriendCommand;
import com.innerfriends.friends.domain.usecase.EstablishAFriendshipToFriendWithFromFriendUseCase;
import com.innerfriends.friends.infrastructure.InstantProvider;
import com.innerfriends.friends.infrastructure.outbox.ToFriendEstablishedAFriendshipWithFromFriendEvent;
import io.debezium.outbox.quarkus.ExportedEvent;

import javax.enterprise.context.ApplicationScoped;
import javax.enterprise.event.Event;
import javax.transaction.Transactional;
import java.util.Objects;

@ApplicationScoped
public class ManagedEstablishAFriendshipToFriendWithFromFriendUseCase implements UseCase<Friend, EstablishAFriendshipToFriendWithFromFriendCommand> {

    private final EstablishAFriendshipToFriendWithFromFriendUseCase establishAFriendshipToFriendWithFromFriendUseCase;
    private final Event<ExportedEvent<?, ?>> event;
    private final ObjectMapper objectMapper;
    private final InstantProvider instantProvider;

    public ManagedEstablishAFriendshipToFriendWithFromFriendUseCase(final EstablishAFriendshipToFriendWithFromFriendUseCase establishAFriendshipToFriendWithFromFriendUseCase,
                                                                    final Event<ExportedEvent<?, ?>> event,
                                                                    final ObjectMapper objectMapper,
                                                                    final InstantProvider instantProvider) {
        this.establishAFriendshipToFriendWithFromFriendUseCase = Objects.requireNonNull(establishAFriendshipToFriendWithFromFriendUseCase);
        this.event = Objects.requireNonNull(event);
        this.objectMapper = Objects.requireNonNull(objectMapper);
        this.instantProvider = Objects.requireNonNull(instantProvider);
    }

    @Transactional
    @Override
    public Friend execute(final EstablishAFriendshipToFriendWithFromFriendCommand command) {
        final Friend friendConnected = establishAFriendshipToFriendWithFromFriendUseCase.execute(command);
        event.fire(ToFriendEstablishedAFriendshipWithFromFriendEvent.of(friendConnected, objectMapper, instantProvider));
        return friendConnected;
    }
}
