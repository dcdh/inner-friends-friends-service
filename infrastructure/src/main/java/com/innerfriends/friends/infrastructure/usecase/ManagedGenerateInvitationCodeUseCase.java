package com.innerfriends.friends.infrastructure.usecase;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.innerfriends.friends.domain.InvitationCodeGenerated;
import com.innerfriends.friends.domain.UseCase;
import com.innerfriends.friends.domain.usecase.GenerateInvitationCodeCommand;
import com.innerfriends.friends.domain.usecase.GenerateInvitationCodeUseCase;
import com.innerfriends.friends.infrastructure.InstantProvider;
import com.innerfriends.friends.infrastructure.outbox.InvitationCodeGeneratedEvent;
import io.debezium.outbox.quarkus.ExportedEvent;

import javax.enterprise.context.ApplicationScoped;
import javax.enterprise.event.Event;
import javax.transaction.Transactional;
import java.util.Objects;

@ApplicationScoped
public class ManagedGenerateInvitationCodeUseCase implements UseCase<InvitationCodeGenerated, GenerateInvitationCodeCommand> {

    private final GenerateInvitationCodeUseCase generateInvitationCodeUseCase;
    private final Event<ExportedEvent<?, ?>> event;
    private final ObjectMapper objectMapper;
    private final InstantProvider instantProvider;

    public ManagedGenerateInvitationCodeUseCase(final GenerateInvitationCodeUseCase generateInvitationCodeUseCase,
                                                final Event<ExportedEvent<?, ?>> event,
                                                final ObjectMapper objectMapper,
                                                final InstantProvider instantProvider) {
        this.generateInvitationCodeUseCase = Objects.requireNonNull(generateInvitationCodeUseCase);
        this.event = Objects.requireNonNull(event);
        this.objectMapper = Objects.requireNonNull(objectMapper);
        this.instantProvider = Objects.requireNonNull(instantProvider);
    }

    @Transactional
    @Override
    public InvitationCodeGenerated execute(final GenerateInvitationCodeCommand command) {
        final InvitationCodeGenerated invitationCodeGenerated = generateInvitationCodeUseCase.execute(command);
        event.fire(InvitationCodeGeneratedEvent.of(invitationCodeGenerated, objectMapper, instantProvider));
        return invitationCodeGenerated;
    }
}
