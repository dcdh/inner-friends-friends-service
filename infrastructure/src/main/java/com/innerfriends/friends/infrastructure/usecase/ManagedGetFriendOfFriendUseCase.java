package com.innerfriends.friends.infrastructure.usecase;

import com.innerfriends.friends.domain.FriendOfFriend;
import com.innerfriends.friends.domain.UseCase;
import com.innerfriends.friends.domain.usecase.GetFriendOfFriendCommand;
import com.innerfriends.friends.domain.usecase.GetFriendOfFriendUseCase;

import javax.enterprise.context.ApplicationScoped;
import javax.transaction.Transactional;
import java.util.Objects;

@ApplicationScoped
public class ManagedGetFriendOfFriendUseCase implements UseCase<FriendOfFriend, GetFriendOfFriendCommand> {

    private GetFriendOfFriendUseCase getFriendOfFriendUseCase;

    public ManagedGetFriendOfFriendUseCase(final GetFriendOfFriendUseCase getFriendOfFriendUseCase) {
        this.getFriendOfFriendUseCase = Objects.requireNonNull(getFriendOfFriendUseCase);
    }

    @Transactional
    @Override
    public FriendOfFriend execute(final GetFriendOfFriendCommand command) {
        return getFriendOfFriendUseCase.execute(command);
    }

}
