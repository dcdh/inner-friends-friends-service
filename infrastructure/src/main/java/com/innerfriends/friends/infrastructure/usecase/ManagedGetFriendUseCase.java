package com.innerfriends.friends.infrastructure.usecase;

import com.innerfriends.friends.domain.Friend;
import com.innerfriends.friends.domain.UseCase;
import com.innerfriends.friends.domain.usecase.GetFriendCommand;
import com.innerfriends.friends.domain.usecase.GetFriendUseCase;

import javax.enterprise.context.ApplicationScoped;
import javax.transaction.Transactional;
import java.util.Objects;

@ApplicationScoped
public class ManagedGetFriendUseCase implements UseCase<Friend, GetFriendCommand> {

    private final GetFriendUseCase getFriendUseCase;

    public ManagedGetFriendUseCase(final GetFriendUseCase getFriendUseCase) {
        this.getFriendUseCase = Objects.requireNonNull(getFriendUseCase);
    }

    @Transactional
    @Override
    public Friend execute(final GetFriendCommand command) {
        return getFriendUseCase.execute(command);
    }

}
