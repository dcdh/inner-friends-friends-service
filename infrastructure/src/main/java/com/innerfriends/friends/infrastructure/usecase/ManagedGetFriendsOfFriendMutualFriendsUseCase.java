package com.innerfriends.friends.infrastructure.usecase;

import com.innerfriends.friends.domain.FriendsOfFriendMutualFriends;
import com.innerfriends.friends.domain.UseCase;
import com.innerfriends.friends.domain.usecase.GetFriendsOfFriendMutualFriendsCommand;
import com.innerfriends.friends.domain.usecase.GetFriendsOfFriendMutualFriendsUseCase;

import javax.enterprise.context.ApplicationScoped;
import javax.transaction.Transactional;
import java.util.Objects;

@ApplicationScoped
public class ManagedGetFriendsOfFriendMutualFriendsUseCase implements UseCase<FriendsOfFriendMutualFriends, GetFriendsOfFriendMutualFriendsCommand> {

    private GetFriendsOfFriendMutualFriendsUseCase getFriendsOfFriendMutualFriendsUseCase;

    public ManagedGetFriendsOfFriendMutualFriendsUseCase(final GetFriendsOfFriendMutualFriendsUseCase getFriendsOfFriendMutualFriendsUseCase) {
        this.getFriendsOfFriendMutualFriendsUseCase = Objects.requireNonNull(getFriendsOfFriendMutualFriendsUseCase);
    }

    @Transactional
    @Override
    public FriendsOfFriendMutualFriends execute(final GetFriendsOfFriendMutualFriendsCommand command) {
        return getFriendsOfFriendMutualFriendsUseCase.execute(command);
    }

}
