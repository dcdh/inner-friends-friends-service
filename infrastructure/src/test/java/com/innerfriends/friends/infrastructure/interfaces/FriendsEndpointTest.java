package com.innerfriends.friends.infrastructure.interfaces;

import com.innerfriends.friends.domain.*;
import com.innerfriends.friends.domain.usecase.*;
import com.innerfriends.friends.infrastructure.usecase.*;
import io.quarkus.test.junit.QuarkusTest;
import io.quarkus.test.junit.mockito.InjectMock;
import io.quarkus.test.security.TestSecurity;
import io.quarkus.test.security.jwt.Claim;
import io.quarkus.test.security.jwt.JwtSecurity;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.junit.jupiter.MockitoExtension;

import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;

import static io.restassured.RestAssured.given;
import static org.approvaltests.Approvals.verifyJson;
import static org.hamcrest.Matchers.equalTo;
import static org.mockito.Mockito.doReturn;
import static org.mockito.Mockito.doThrow;

@QuarkusTest
@ExtendWith(MockitoExtension.class)
public class FriendsEndpointTest {

    @InjectMock
    private ManagedGetFriendUseCase managedGetFriendUseCase;

    @InjectMock
    private ManagedGetInFriendshipWithUseCase managedGetInFriendshipWithUseCase;

    @InjectMock
    private ManagedGenerateInvitationCodeUseCase managedGenerateInvitationCodeUseCase;

    @InjectMock
    private ManagedWriteBioUseCase managedWriteBioUseCase;

    @InjectMock
    private ManagedEstablishAFriendshipToFriendWithFromFriendUseCase managedEstablishAFriendshipToFriendWithFromFriendUseCase;

    @InjectMock
    private ManagedGetFriendOfFriendUseCase managedGetFriendOfFriendUseCase;

    @InjectMock
    private ManagedGetInFriendshipWithMutualFriendsUseCase managedGetInFriendshipWithMutualFriendsUseCase;

    @InjectMock
    private ManagedGetFriendsOfFriendMutualFriendsUseCase managedGetFriendsOfFriendMutualFriendsUseCase;

    @InjectMock
    private GetFriendMayKnowUseCase getFriendMayKnowUseCase;

    @InjectMock
    private ListFriendMayKnowUseCase listFriendMayKnowUseCase;

    @Test
    @TestSecurity(user = "mario", roles = "friend")
    @JwtSecurity(claims = {
            @Claim(key = "pseudo", value = "Mario")
    })
    public void should_get_friend() {
        // Given
        doReturn(new Friend(new FriendId("Mario")))
                .when(managedGetFriendUseCase)
                .execute(new GetFriendCommand(new FriendId("Mario")));

        // When && Then
        final String jsonResponse = given()
                .when()
                .get("/friends/Mario")
                .then()
                .log().all()
                .statusCode(200)
                .extract().response().body().prettyPrint();
        verifyJson(jsonResponse);
    }

    @Test
    @TestSecurity(user = "mario", roles = "friend")
    @JwtSecurity(claims = {
            @Claim(key = "pseudo", value = "Mario")
    })
    public void should_get_friend_fail_when_friend_is_unknown() {
        // Given
        doThrow(new FriendUnknownException(new FriendId("Mario")))
                .when(managedGetFriendUseCase)
                .execute(new GetFriendCommand(new FriendId("Mario")));

        // When && Then
        given()
                .when()
                .get("/friends/Mario")
                .then()
                .log().all()
                .statusCode(404);
    }

    @Test
    @TestSecurity(user = "mario", roles = "friend")
    @JwtSecurity(claims = {
            @Claim(key = "pseudo", value = "Mario")
    })
    public void should_generate_invitation_code() {
        // Given
        doReturn(new InvitationCodeGenerated(
                new FromFriendId("Mario"),
                new InvitationCode(new UUID(0, 0)),
                new GeneratedAt(LocalDateTime.of(2021, 11, 29, 0, 0))))
                .when(managedGenerateInvitationCodeUseCase)
                .execute(new GenerateInvitationCodeCommand(new FromFriendId("Mario")));

        // When && Then
        final String jsonResponse = given()
                .when()
                .post("/friends/generateInvitationCode")
                .then()
                .log().all()
                .statusCode(200)
                .extract().response().body().prettyPrint();
        verifyJson(jsonResponse);
    }

    @Test
    @TestSecurity(user = "mario", roles = "friend")
    @JwtSecurity(claims = {
            @Claim(key = "pseudo", value = "Mario")
    })
    public void should_write_bio() {
        // Given
        doReturn(new Friend(new FriendId("Mario")).writeBio(new Bio("super plumber"), new ExecutedBy("Mario")))
                .when(managedWriteBioUseCase)
                .execute(new WriteBioCommand(new FriendId("Mario"), new Bio("super plumber"), new ExecutedBy("Mario")));

        // When && Then
        final String jsonResponse = given()
                .param("bio", "super plumber")
                .when()
                .post("/friends/writeBio")
                .then()
                .log().all()
                .statusCode(200)
                .extract().response().body().prettyPrint();
        verifyJson(jsonResponse);
    }

    @Test
    @TestSecurity(user = "peach", roles = "friend")
    @JwtSecurity(claims = {
            @Claim(key = "pseudo", value = "Peach")
    })
    public void should_establish_a_friendship_to_friend_with_from_friend() {
        // Given
        doReturn(new Friend(new FriendId("Mario")).establishAFriendship(new FriendId("Peach")))
                .when(managedEstablishAFriendshipToFriendWithFromFriendUseCase)
                .execute(new EstablishAFriendshipToFriendWithFromFriendCommand(
                        new ToFriendId("Peach"),
                        new InvitationCode(new UUID(0, 0)),
                        new ExecutedBy("Peach")));

        // When && Then
        final String jsonResponse = given()
                .param("invitationCode", "00000000-0000-0000-0000-000000000000")
                .when()
                .post("/friends/establishAFriendship")
                .then()
                .log().all()
                .statusCode(200)
                .extract().response().body().prettyPrint();
        verifyJson(jsonResponse);
    }

    @Test
    @TestSecurity(user = "peach", roles = "friend")
    @JwtSecurity(claims = {
            @Claim(key = "pseudo", value = "Peach")
    })
    public void should_establish_a_friendship_to_friend_with_from_friend_fail_when_already_a_friend() {
        // Given
        doThrow(new AlreadyAFriendException())
                .when(managedEstablishAFriendshipToFriendWithFromFriendUseCase)
                .execute(new EstablishAFriendshipToFriendWithFromFriendCommand(
                        new ToFriendId("Peach"),
                        new InvitationCode(new UUID(0, 0)),
                        new ExecutedBy("Peach")));

        // When && Then
        given()
                .param("invitationCode", "00000000-0000-0000-0000-000000000000")
                .when()
                .post("/friends/establishAFriendship")
                .then()
                .log().all()
                .statusCode(400)
                .body(equalTo("Already a friend"));
    }

    @Test
    @TestSecurity(user = "peach", roles = "friend")
    @JwtSecurity(claims = {
            @Claim(key = "pseudo", value = "Peach")
    })
    public void should_establish_a_friendship_to_friend_with_from_friend_fail_when_invitation_code_does_not_exists() {
        // Given
        doThrow(new UnknownInvitationCodeException())
                .when(managedEstablishAFriendshipToFriendWithFromFriendUseCase)
                .execute(new EstablishAFriendshipToFriendWithFromFriendCommand(
                        new ToFriendId("Peach"),
                        new InvitationCode(new UUID(0, 0)),
                        new ExecutedBy("Peach")));

        // When && Then
        given()
                .param("invitationCode", "00000000-0000-0000-0000-000000000000")
                .when()
                .post("/friends/establishAFriendship")
                .then()
                .log().all()
                .statusCode(404)
                .body(equalTo("Unknown invitation code"));
    }

    @Test
    @TestSecurity(user = "peach", roles = "friend")
    @JwtSecurity(claims = {
            @Claim(key = "pseudo", value = "Peach")
    })
    public void should_establish_a_friendship_to_friend_with_from_friend_fail_when_invitation_code_is_invalide() {
        // Given
        doThrow(new InvitationCodeInvalideException())
                .when(managedEstablishAFriendshipToFriendWithFromFriendUseCase)
                .execute(new EstablishAFriendshipToFriendWithFromFriendCommand(
                        new ToFriendId("Peach"),
                        new InvitationCode(new UUID(0, 0)),
                        new ExecutedBy("Peach")));

        // When && Then
        given()
                .param("invitationCode", "00000000-0000-0000-0000-000000000000")
                .param("executedBy", "Peach")
                .when()
                .post("/friends/establishAFriendship")
                .then()
                .log().all()
                .statusCode(400)
                .body(equalTo("Invitation code invalid. Please generate a new one."));
    }

    @Test
    @TestSecurity(user = "mario", roles = "friend")
    @JwtSecurity(claims = {
            @Claim(key = "pseudo", value = "Mario")
    })
    public void should_get_in_friendship_with() {
        // Given
        doReturn(new InFriendshipWith(
                new InFriendshipWithId("DonkeyKong"), List.of(new FriendOfFriendId("Pauline")),
                new Bio(""), new Version(2l)))
                .when(managedGetInFriendshipWithUseCase)
                .execute(new GetInFriendshipWithCommand(
                        new FriendId("Mario"), new InFriendshipWithId("DonkeyKong")));

        // When && Then
        final String jsonResponse = given()
                .when()
                .get("/friends/Mario/inFriendshipsWith/DonkeyKong")
                .then()
                .log().all()
                .statusCode(200)
                .extract().response().body().prettyPrint();
        verifyJson(jsonResponse);
    }

    @Test
    @TestSecurity(user = "mario", roles = "friend")
    @JwtSecurity(claims = {
            @Claim(key = "pseudo", value = "Mario")
    })
    public void should_get_in_friendship_with_fail_when_not_in_friendship() {
        // Given
        doThrow(new NotInFriendshipWithException())
                .when(managedGetInFriendshipWithUseCase)
                .execute(new GetInFriendshipWithCommand(
                        new FriendId("Mario"), new InFriendshipWithId("DonkeyKong")));

        // When && Then
        given()
                .when()
                .get("/friends/Mario/inFriendshipsWith/DonkeyKong")
                .then()
                .log().all()
                .statusCode(400)
                .body(equalTo("Not in friendship with"));
    }

    @Test
    @TestSecurity(user = "mario", roles = "friend")
    @JwtSecurity(claims = {
            @Claim(key = "pseudo", value = "Mario")
    })
    public void should_get_in_friendship_with_mutual_friends() {
        // Given
        doReturn(new InFriendshipWithMutualFriends(
                new FriendId("Mario"), new InFriendshipWithId("DonkeyKong"),
                List.of(new MutualFriendId("Pauline"))))
                .when(managedGetInFriendshipWithMutualFriendsUseCase)
                .execute(new GetInFriendshipWithMutualFriendsCommand(
                        new FriendId("Mario"), new InFriendshipWithId("DonkeyKong")));

        // When && Then
        final String jsonResponse = given()
                .when()
                .get("/friends/Mario/inFriendshipsWith/DonkeyKong/mutualFriends")
                .then()
                .log().all()
                .statusCode(200)
                .extract().response().body().prettyPrint();
        verifyJson(jsonResponse);
    }

    @Test
    @TestSecurity(user = "mario", roles = "friend")
    @JwtSecurity(claims = {
            @Claim(key = "pseudo", value = "Mario")
    })
    public void should_get_friends_of_friend() {
        // Given
        doReturn(new FriendOfFriend(new FriendOfFriendId("Pauline"), new Bio(""), new Version(1l), true))
                .when(managedGetFriendOfFriendUseCase)
                .execute(new GetFriendOfFriendCommand(
                        new FriendId("Mario"), new InFriendshipWithId("DonkeyKong"), new FriendOfFriendId("Pauline")));

        // When && Then
        final String jsonResponse = given()
                .when()
                .get("/friends/Mario/inFriendshipsWith/DonkeyKong/friendsOfFriend/Pauline")
                .then()
                .log().all()
                .statusCode(200)
                .extract().response().body().prettyPrint();
        verifyJson(jsonResponse);
    }

    @Test
    @TestSecurity(user = "mario", roles = "friend")
    @JwtSecurity(claims = {
            @Claim(key = "pseudo", value = "Mario")
    })
    public void should_get_friends_of_friend_when_not_in_friendship_with() {
        // Given
        doThrow(new NotInFriendshipWithException())
                .when(managedGetFriendOfFriendUseCase)
                .execute(new GetFriendOfFriendCommand(
                        new FriendId("Mario"), new InFriendshipWithId("DonkeyKong"), new FriendOfFriendId("Pauline")));

        // When && Then
        given()
                .when()
                .get("/friends/Mario/inFriendshipsWith/DonkeyKong/friendsOfFriend/Pauline")
                .then()
                .log().all()
                .statusCode(400)
                .body(equalTo("Not in friendship with"));
    }

    @Test
    @TestSecurity(user = "mario", roles = "friend")
    @JwtSecurity(claims = {
            @Claim(key = "pseudo", value = "Mario")
    })
    public void should_get_friends_of_friend_when_not_a_friend_of_friend() {
        // Given
        doThrow(new NotAFriendOfFriendException())
                .when(managedGetFriendOfFriendUseCase)
                .execute(new GetFriendOfFriendCommand(
                        new FriendId("Mario"), new InFriendshipWithId("DonkeyKong"), new FriendOfFriendId("Pauline")));

        // When && Then
        given()
                .when()
                .get("/friends/Mario/inFriendshipsWith/DonkeyKong/friendsOfFriend/Pauline")
                .then()
                .log().all()
                .statusCode(400)
                .body(equalTo("Not a friend of friend"));
    }

    @Test
    @TestSecurity(user = "mario", roles = "friend")
    @JwtSecurity(claims = {
            @Claim(key = "pseudo", value = "Mario")
    })
    public void should_get_friends_of_friend_mutual_friends() {
        // Given
        doReturn(new FriendsOfFriendMutualFriends(
                new FriendId("Mario"), new InFriendshipWithId("DonkeyKong"), new FriendOfFriendId("Pauline"),
                List.of(new MutualFriendId("Luigi"))))
                .when(managedGetFriendsOfFriendMutualFriendsUseCase)
                .execute(new GetFriendsOfFriendMutualFriendsCommand(
                        new FriendId("Mario"), new InFriendshipWithId("DonkeyKong"), new FriendOfFriendId("Pauline")));

        // When && Then
        final String jsonResponse = given()
                .when()
                .get("/friends/Mario/inFriendshipsWith/DonkeyKong/friendsOfFriend/Pauline/mutualFriends")
                .then()
                .log().all()
                .statusCode(200)
                .extract().response().body().prettyPrint();
        verifyJson(jsonResponse);
    }

    @Test
    @TestSecurity(user = "mario", roles = "friend")
    @JwtSecurity(claims = {
            @Claim(key = "pseudo", value = "Mario")
    })
    public void should_get_friend_may_know() {
        // Given
        doReturn(new FriendMayKnow(
                new FriendMayKnowId("Peach"),
                new Bio("I am a princess"),
                new Version(1l),
                List.of(new MutualFriendId("Luigi"))
        ))
                .when(getFriendMayKnowUseCase)
                .execute(new GetFriendMayKnowCommand(new FriendId("Mario"), new FriendMayKnowId("Peach")));

        // When && Then
        final String jsonResponse = given()
                .when()
                .get("/friends/mayKnow/Peach")
                .then()
                .log().all()
                .statusCode(200)
                .extract().response().body().prettyPrint();
        verifyJson(jsonResponse);
    }

    @Test
    @TestSecurity(user = "mario", roles = "friend")
    @JwtSecurity(claims = {
            @Claim(key = "pseudo", value = "Mario")
    })
    public void should_list_friends_may_know() {
        // Given
        doReturn(List.of(new FriendMayKnowId("Peach")))
                .when(listFriendMayKnowUseCase).execute(new ListFriendMayKnowCommand(new FriendId("Mario"), 2l));

        // When && Then
        final String jsonResponse = given()
                .param("nbOf", 2l)
                .when()
                .post("/friends/mayKnow")
                .then()
                .log().all()
                .statusCode(200)
                .extract().response().body().prettyPrint();
        verifyJson(jsonResponse);
    }

}
