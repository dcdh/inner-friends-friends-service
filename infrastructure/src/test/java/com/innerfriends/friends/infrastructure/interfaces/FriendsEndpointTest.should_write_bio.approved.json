{
  "pseudo": "Mario",
  "bio": "super plumber",
  "friends": [
    "DamDamDeo"
  ],
  "nbOfFriends": 1,
  "version": 1
}